ElectricBillApi/Controllers/ElectricProviderController.cs

```diff
@@ -0,0 +1,77 @@
+﻿using ElectricBillApi.Exceptions;
+using ElectricBillApi.Interfaces;
+using ElectricBillApi.Models;
+using ElectricBillApi.Services;
+using Microsoft.AspNetCore.Mvc;
+
+namespace ElectricBillApi.Controllers
+{
+    [ApiController]
+    [Route("api/[controller]")]
+    public class ElectricProviderController : ControllerBase
+    {
+        private readonly IElectricProviderPicker _electricProviderPicker;
+
+        public ElectricProviderController(IElectricProviderPicker electricProviderPicker)
+        {
+            _electricProviderPicker = electricProviderPicker;
+        }
+
+        [HttpGet]
+        public IActionResult GetBestProvider(Address address)
+        {
+            try
+            {
+                var provider = _electricProviderPicker.FindCheapest(address);
+
+                return Ok($"{provider.Name}, {provider.CalculatePrice(address)}, {provider.CurrentSubscribedPlant}");
+            }
+            catch (NoSubscribedPowerPlantException ex)
+            {
+                return NotFound(ex.Message);
+            }
+            catch (NoValidAdressException ex)
```

Comment:
If address involves validation, it's better to just allow 500 to be returned- it happened in the server and it actually did some work.
400 implies that we didn't even start processing the request- the shape of request was not understood.

Reply:
Ok. I removed the exception and returned 500 if provider is null.